crates/feldera-storage/src/file/reader.rs

```diff
@@ -0,0 +1,1871 @@
+    n_rows: u64,
+}
+
+impl Column {
+    fn new(info: &FileTrailerColumn) -> Result<Self, Error> {
+        let FileTrailerColumn {
+            index_offset,
+            index_size,
+            n_rows,
+        } = *info;
+        let root = if n_rows != 0 {
+            let location = match BlockLocation::new(index_offset, index_size as usize) {
+                Ok(location) => location,
+                Err(_) => {
+                    return Err(Error::Corruption(CorruptionError::InvalidIndexRoot {
+                        index_offset,
+                        index_size,
+                    }))
+                }
+            };
+            Some(TreeNode {
+                location,
+                node_type: NodeType::Index,
+                depth: 0,
+                rows: 0..n_rows,
+            })
+        } else {
+            None
+        };
+        Ok(Self { root, n_rows })
+    }
+    fn empty() -> Self {
+        Self {
+            root: None,
+            n_rows: 0,
+        }
+    }
+}
+
+struct ReaderInner<T> {
+    file: File,
+    columns: Vec<Column>,
+
+    /// `fn() -> T` is `Send` and `Sync` regardless of `T`.  See
+    /// <https://doc.rust-lang.org/nomicon/phantom-data.html>.
+    _phantom: PhantomData<fn() -> T>,
+}
+
+fn read_block(file: &File, location: BlockLocation) -> Result<AlignedVec, Error> {
+    let mut block = AlignedVec::with_capacity(location.size);
+    // XXX This zeros the whole buffer before reading into it.
+    block.resize(location.size, 0);
+    file.read_exact_at(block.as_mut_slice(), location.offset)?;
+    let computed_checksum = crc32c(&block[4..]);
+    let checksum = u32::from_le_bytes(block[..4].try_into().unwrap());
+    if checksum != computed_checksum {
+        let BlockLocation { size, offset } = location;
+        return Err(CorruptionError::InvalidChecksum {
+            size,
+            offset,
+            checksum,
+            computed_checksum,
+        }
+        .into());
+    }
+    Ok(block)
+}
+
+/// Layer file column specification.
+///
+/// A column specification must take the form `K1, A1, N1`, where `(K1, A1)` is
+/// the first column's key and auxiliary data types.  If there is only one
+/// column, `N1` is `()`; otherwise, it is `(K2, A2, N2)`, where `(K2, A2)` is
+/// the second column's key and auxiliary data types.  If there are only two
+/// columns, `N2` is `()`, otherwise it is `(K3, A3, N3)`; and so on.  Thus:
+///
+/// * For one column, `T` is `(K1, A1, ())`.
+///
+/// * For two columns, `T` is `(K1, A1, (K2, A2, ()))`.
+///
+/// * For three columns, `T` is `(K1, A1, (K2, A2, (K3, A3, ())))`.
+pub trait ColumnSpec {
+    /// Returns the number of columns in this `ColumnSpec`.
+    fn n_columns() -> usize;
+}
+impl ColumnSpec for () {
+    fn n_columns() -> usize {
+        0
+    }
+}
+impl<K, A, N> ColumnSpec for (K, A, N)
+where
+    K: Rkyv,
+    A: Rkyv,
+    N: ColumnSpec,
+{
+    fn n_columns() -> usize {
+        1 + N::n_columns()
```

Comment:
this is a new one for me.
Didn't know you can have recursive traits

Reply:
I was pretty amazed that it worked!